File: clients/python/prismsearch/src/prismsearch/query.py
"""Pipe-style query builder for Prism searches."""

from __future__ import annotations
from typing import Any, TYPE_CHECKING

if TYPE_CHECKING:
    from prismsearch.client import Prismsearch
    from prismsearch.models import SearchResults


class Query:
    """Chainable query builder.

    Usage::

        results = (
            Query("products", "wireless headphones")
            .fields(["title", "description"])
            .limit(20)
            .highlight(fields=["title"])
            .execute(client)
        )
    """

    def __init__(self, collection: str, query: str | None = None):
        self.collection = collection
        self._query = query
        self._vector: list[float] | None = None
        self._fields: list[str] = []
        self._limit: int = 10
        self._offset: int = 0
        self._merge_strategy: str | None = None
        self._text_weight: float | None = None
        self._vector_weight: float | None = None
        self._highlight: dict | None = None
        self._rerank: dict | None = None
        self._min_score: float | None = None
        self._score_function: str | None = None
        self._rrf_k: int | None = None
        self._aggregations: list[dict] = []

    def fields(self, fields: list[str]) -> Query:
        self._fields = fields
        return self

    def limit(self, n: int) -> Query:
        self._limit = n
        return self

    def offset(self, n: int) -> Query:
        self._offset = n
        return self

    def vector(self, vec: list[float]) -> Query:
        self._vector = vec
        return self

    def min_score(self, s: float) -> Query:
        self._min_score = s
        return self

    def score_function(self, expr: str) -> Query:
        self._score_function = expr
        return self

    def merge_strategy(self, s: str) -> Query:
        self._merge_strategy = s
        return self

    def text_weight(self, w: float) -> Query:
        self._text_weight = w
        return self

    def vector_weight(self, w: float) -> Query:
        self._vector_weight = w
        return self

    def rrf_k(self, k: int) -> Query:
        self._rrf_k = k
        return self

    def highlight(self, *, fields: list[str], pre_tag: str = "<em>", post_tag: str = "</em>", fragment_size: int = 150, number_of_fragments: int = 3) -> Query:
        self._highlight = {
            "fields": fields,
            "pre_tag": pre_tag,
            "post_tag": post_tag,
            "fragment_size": fragment_size,
            "number_of_fragments": number_of_fragments,
        }
        return self

    def aggregate(self, name: str, **kwargs: Any) -> Query:
        agg = {"name": name, **kwargs}
        self._aggregations.append(agg)
        return self

    def to_request_body(self) -> dict[str, Any]:
        body: dict[str, Any] = {"limit": self._limit}
        if self._query is not None:
            body["query"] = self._query
        if self._vector is not None:
            body["vector"] = self._vector
        if self._fields:
            body["fields"] = self._fields
        if self._offset > 0:
            body["offset"] = self._offset
        if self._merge_strategy:
            body["merge_strategy"] = self._merge_strategy
        if self._text_weight is not None:
            body["text_weight"] = self._text_weight
        if self._vector_weight is not None:
            body["vector_weight"] = self._vector_weight
        if self._highlight:
            body["highlight"] = self._highlight
        if self._rerank:
            body["rerank"] = self._rerank
        if self._min_score is not None:
            body["min_score"] = self._min_score
        if self._score_function:
            body["score_function"] = self._score_function
        if self._rrf_k is not None:
            body["rrf_k"] = self._rrf_k
        return body

    def to_aggregate_body(self) -> dict[str, Any]:
        body: dict[str, Any] = {
            "aggregations": self._aggregations,
            "scan_limit": self._limit,
        }
        if self._query:
            body["query"] = self._query
        return body

    def execute(self, client: Prismsearch) -> SearchResults:
        """Execute search and return typed results."""
        return client.search(self.collection, self.to_request_body())

    def execute_aggs(self, client: Prismsearch) -> dict:
        """Execute aggregation query."""
        return client.aggregate(self.collection, self.to_aggregate_body())

File: clients/elixir/prismsearch/lib/prismsearch/query.ex
defmodule Prismsearch.Query do
  @moduledoc """
  Pipe-friendly query builder for Prism searches.

  ## Example

      "products"
      |> Query.new("wireless headphones")
      |> Query.fields(["title", "description"])
      |> Query.limit(20)
      |> Query.highlight(fields: ["title"])
      |> Prismsearch.search(client)
  """

  defstruct [
    :collection,
    :query,
    :vector,
    :merge_strategy,
    :text_weight,
    :vector_weight,
    :highlight,
    :rerank,
    :min_score,
    :score_function,
    :rrf_k,
    fields: [],
    limit: 10,
    offset: 0,
    aggregations: []
  ]

  @type t :: %__MODULE__{}

  @doc "Create a new query for a collection."
  def new(collection, query \\ nil) do
    %__MODULE__{collection: collection, query: query}
  end

  def fields(%__MODULE__{} = q, fields) when is_list(fields), do: %{q | fields: fields}
  def limit(%__MODULE__{} = q, n) when is_integer(n), do: %{q | limit: n}
  def offset(%__MODULE__{} = q, n) when is_integer(n), do: %{q | offset: n}
  def min_score(%__MODULE__{} = q, s) when is_number(s), do: %{q | min_score: s}
  def score_function(%__MODULE__{} = q, expr) when is_binary(expr), do: %{q | score_function: expr}
  def vector(%__MODULE__{} = q, vec) when is_list(vec), do: %{q | vector: vec}
  def merge_strategy(%__MODULE__{} = q, s), do: %{q | merge_strategy: s}
  def text_weight(%__MODULE__{} = q, w), do: %{q | text_weight: w}
  def vector_weight(%__MODULE__{} = q, w), do: %{q | vector_weight: w}
  def rrf_k(%__MODULE__{} = q, k) when is_integer(k), do: %{q | rrf_k: k}

  @doc "Set highlight configuration."
  def highlight(%__MODULE__{} = q, opts) do
    h =
      opts
      |> Enum.into(%{}, fn {k, v} -> {to_string(k), v} end)

    %{q | highlight: h}
  end

  @doc "Add an aggregation to the query."
  def aggregate(%__MODULE__{} = q, name, opts) do
    agg =
      opts
      |> Enum.into(%{}, fn {k, v} -> {to_string(k), v} end)
      |> Map.put("name", name)

    %{q | aggregations: q.aggregations ++ [agg]}
  end

  @doc "Convert query to request body map (without collection)."
  def to_request_body(%__MODULE__{} = q) do
    %{}
    |> maybe_put("query", q.query)
    |> maybe_put("vector", q.vector)
    |> maybe_put_list("fields", q.fields)
    |> Map.put("limit", q.limit)
    |> maybe_put_nonzero("offset", q.offset)
    |> maybe_put("merge_strategy", q.merge_strategy)
    |> maybe_put("text_weight", q.text_weight)
    |> maybe_put("vector_weight", q.vector_weight)
    |> maybe_put("highlight", q.highlight)
    |> maybe_put("rerank", q.rerank)
    |> maybe_put("min_score", q.min_score)
    |> maybe_put("score_function", q.score_function)
    |> maybe_put("rrf_k", q.rrf_k)
  end

  @doc "Convert query to aggregate request body."
  def to_aggregate_body(%__MODULE__{} = q) do
    %{"aggregations" => q.aggregations}
    |> maybe_put("query", q.query)
    |> Map.put("scan_limit", q.limit)
  end

  defp maybe_put(map, _key, nil), do: map
  defp maybe_put(map, key, value), do: Map.put(map, key, value)

  defp maybe_put_list(map, _key, []), do: map
  defp maybe_put_list(map, key, list), do: Map.put(map, key, list)

  defp maybe_put_nonzero(map, _key, 0), do: map
  defp maybe_put_nonzero(map, key, value), do: Map.put(map, key, value)
end
